zebra-chain/src/block/tests.rs

```diff
@@ -130,6 +136,160 @@ fn deserialize_block() {
+    let mut data = Vec::new();
+    block
+        .zcash_serialize(&mut data)
+        .expect("block should serialize as we are not limiting generation yet");
+
+    // Deserialize will now fail
+    Block::zcash_deserialize(&data[..]).expect_err("block should not deserialize");
+}
+
+#[test]
+fn block_limits_single_tx() {
+    // Test block limit with a big single transaction
+    // Dummy input and output
+    let input =
+        TransparentInput::zcash_deserialize(&zebra_test::vectors::DUMMY_INPUT1[..]).unwrap();
+    let output =
+        TransparentOutput::zcash_deserialize(&zebra_test::vectors::DUMMY_OUTPUT1[..]).unwrap();
+
+    // A block header
+    let blockheader = generate_block_header();
+
+    // Serialize header
+    let mut data_header = Vec::new();
+    blockheader
+        .zcash_serialize(&mut data_header)
+        .expect("Block header should serialize");
+
+    // Serialize a LockTime
+    let locktime = LockTime::Time(DateTime::<Utc>::from_utc(
+        NaiveDateTime::from_timestamp(61, 0),
+        Utc,
+    ));
+    let mut data_locktime = Vec::new();
+    locktime
+        .zcash_serialize(&mut data_locktime)
+        .expect("LockTime should serialize");
+
+    // Calculate the number of inputs we need
+    let mut max_inputs_in_tx = (MAX_BLOCK_BYTES as usize
+        - data_header.len()
+        - *&zebra_test::vectors::DUMMY_OUTPUT1[..].len()
+        - data_locktime.len())
+        / (*&zebra_test::vectors::DUMMY_INPUT1[..].len() - 1);
+
+    let mut outputs = Vec::new();
+
+    // Create inputs to be just below the limit
+    let mut inputs = std::iter::repeat(input.clone())
+        .take(max_inputs_in_tx)
+        .collect::<Vec<_>>();
+
+    // 1 single output
+    outputs.push(output);
+
+    // Create a big transaction
+    let big_transaction = Transaction::V1 {
+        inputs: inputs.clone(),
+        outputs: outputs.clone(),
+        lock_time: locktime,
+    };
+
+    // Put the big transaction into a block
+    let transactions = vec![Arc::new(big_transaction.clone())];
+    let block = Block {
+        header: blockheader,
+        transactions: transactions.clone(),
+    };
+
+    let mut data = Vec::new();
+    block
+        .zcash_serialize(&mut data)
+        .expect("block should serialize as we are not limiting generation yet");
+
+    Block::zcash_deserialize(&data[..])
+        .expect("block should deserialize as we are just below limit");
+
+    // Add 1 more input to the transaction, limit will be reached
+    max_inputs_in_tx = max_inputs_in_tx + 1;
+    inputs.push(input);
+    assert_eq!(max_inputs_in_tx, inputs.len());
+
+    let new_big_transaction = Transaction::V1 {
+        inputs: inputs.clone(),
+        outputs: outputs.clone(),
+        lock_time: locktime,
+    };
+
+    let mut transactions = Vec::new();
+    transactions.push(Arc::new(new_big_transaction.clone()));
+    let new_block = Block {
+        header: blockheader,
+        transactions: transactions,
+    };
```

Comment:
Suggestion: Can you move this large block generation code to `oversize_single_transaction_block()`  in a `zebra_chain::test::generate` module?